lib/OpenQA/Task/Git/Clone.pm

```diff
@@ -83,35 +83,67 @@ sub _git_fetch ($path, $branch_arg) {
+        # dirty git status
+        return "diff-index: $r->{stdout}" if $r->{return_code} >> 8 == 1;
+        die "Failed to diff-index '$path': $r->{stderr}";
+    }
+    return 0;
+}
+
 sub _git_reset_hard ($path, $branch) {
     my $r = run_cmd_with_log_return_error(['git', '-C', $path, 'reset', '--hard', "origin/$branch"]);
     die "Failed to reset to 'origin/$branch': $r->{stderr}" unless $r->{status};
 }
 
 sub _git_clone ($job, $ctx, $path, $url) {
-    $ctx->debug(qq{Updating $path to $url});
-    $url = Mojo::URL->new($url);
-    my $requested_branch = $url->fragment;
-    $url->fragment(undef);
-    my $remote_default_branch = _get_remote_default_branch($url);
-    $requested_branch ||= $remote_default_branch;
-    $ctx->debug(qq{Remote default branch $remote_default_branch});
-    die "Unable to detect remote default branch for '$url'" unless $remote_default_branch;
-
-    if (!-d $path) {
-        _git_clone_url_to_path($url, $path);
-        # update local branch to latest remote branch version
-        _git_fetch($path, "$requested_branch:$requested_branch")
-          if ($requested_branch ne $remote_default_branch);
+    my $requested_branch = '';
+    if ($url) {
+        # given a git url, do initial clone
+        $ctx->debug(qq{Updating $path to $url});
+        $url = Mojo::URL->new($url);
+        $requested_branch = $url->fragment;
+        $url->fragment(undef);
+        my $remote_default_branch = _get_remote_default_branch($url);
+        $requested_branch ||= $remote_default_branch;
+        $ctx->debug(qq{Remote default branch $remote_default_branch});
+        die "Unable to detect remote default branch for '$url'" unless $remote_default_branch;
+
+        if (!-d $path) {
+            _git_clone_url_to_path($url, $path);
+            # update local branch to latest remote branch version
+            _git_fetch($path, "$requested_branch:$requested_branch")
+              if ($requested_branch ne $remote_default_branch);
+        }
+
+        my $origin_url = _git_get_origin_url($path);
+        if ($url ne $origin_url) {
+            $ctx->warn("Local checkout at $path has origin $origin_url but requesting to clone from $url");
+            return;
+        }
     }
-
-    my $origin_url = _git_get_origin_url($path);
-    if ($url ne $origin_url) {
-        $ctx->warn("Local checkout at $path has origin $origin_url but requesting to clone from $url");
+    else {
+        # look for existing clone in the given path
+        my $origin_url = _git_get_origin_url($path);
+        $ctx->debug(qq{Updating $path to $origin_url});
+        $requested_branch = _get_remote_default_branch($origin_url);
+    }
+    if (my $is_dirty = _git_diff_index($path)) {
+        $ctx->warn("Clone at '$path' has dirty status, not updating: $is_dirty");
         return;
     }
 
+    _git_gc($path);
     my $current_branch = _get_current_branch($path);
+    # TODO deal with conflicts because of save_needle
+    # see fetchneedles
```

Comment:
Maybe we can disable running fetchneedles and periodically enqueue a Minion job that runs your new code. Then we wouldn't have to worry about a conflict.

The `save_needle` task could simply try to lock the same Minion locks as this task to avoid conflicts. Note that the same counts for the `delete_needles` task.

Reply:
As the git operation locking is pretty good I would prefer to rely on that so actually call git commands and let them handle the conflicts. WDYT?

Reply:
ok, synchronizing on the complete sequence of commands with a minion lock sounds good then

Reply:
This is now done already by adding a minion guard for `limit_needle_task` in `git_clone`